File: models/Task.ts
import * as Iridium from "iridium";
import {ProjectSummaryDoc, ProjectSummarySchema} from "./Project";
import {ActionSummaryDoc, ActionSummarySchema} from "./Action";

export enum TaskState {
    NotExecuted = 0,
    Executing,
    Failed,
    Passed
}

export interface TaskDoc {
    _id?: string;

    created?: Date;
    executed?: Date;
    completed?: Date;
    
    action: ActionSummaryDoc;
    project: ProjectSummaryDoc;
    
    metadata?: {
        description?: string;
        url?: string;
    };
    
    vars: {
        [name: string]: string;
    };

    state?: TaskState;

    output?: string;
}

@Iridium.Collection("tasks")
export class Task extends Iridium.Instance<TaskDoc, Task> implements TaskDoc {
    @Iridium.ObjectID
    _id: string;

    @Iridium.Property(Date)
    created: Date;
    @Iridium.Property(Date, false)
    executed: Date;
    @Iridium.Property(Date, false)
    completed: Date;
    
    @Iridium.Property({
        description: { $required: false, $type: String },
        url: { $required: false, $type: String }
    })
    metadata: {
        description?: string;
        url?: string;
    };
    
    @Iridium.Property(ActionSummarySchema)
    action: ActionSummaryDoc;

    @Iridium.Property(ProjectSummarySchema)
    project: ProjectSummaryDoc;
    
    @Iridium.Property({
        $propertyType: String
    })
    vars: {
        [name: string]: string;
    };

    @Iridium.Property(Number)
    state: TaskState;

    @Iridium.Property(String)
    output: string;
    
    static onCreating(doc: TaskDoc) {
        doc.created = doc.created || new Date();
        doc.metadata = doc.metadata || {};
        doc.vars = doc.vars || {};
        doc.state = doc.state || TaskState.NotExecuted;
        doc.output = doc.output || "";
    }

    toJSON() {
        return {
            id: this._id,
            created: this.created,
            executed: this.executed,
            completed: this.completed,
            metadata: this.metadata,
            action: this.action,
            project: this.project,
            vars: this.vars,
            state: TaskState[this.state],
            output: this.output
        }
    }
}

File: models/task.go
package models

import (
	"time"

	"gopkg.in/mgo.v2/bson"
)

type TaskState int

const TaskStateNotExecuted = TaskState(0)
const TaskStateExecuting = TaskState(1)
const TaskStateFailed = TaskState(2)
const TaskStatePassed = TaskState(3)

func (t *TaskState) String() string {
	switch *t {
	case TaskStateNotExecuted:
		return "Not Executed"
	case TaskStateExecuting:
		return "Executing"
	case TaskStateFailed:
		return "Failed"
	case TaskStatePassed:
		return "Passed"
	default:
		return "Unknown"
	}
}

type TaskMetadata struct {
	Description string `json:"description"`
	URL         string `json:"url"`
}

type TaskSummary struct {
	ID       bson.ObjectId `json:"id"`
	Metadata TaskMetadata  `json:"metadata"`
}

type Task struct {
	ID       bson.ObjectId `json:"id" bson:"_id"`
	Metadata TaskMetadata  `json:"metadata"`

	Created   time.Time `json:"created"`
	Executed  time.Time `json:"executed"`
	Completed time.Time `json:"completed"`

	Action  *ActionSummary  `json:"action"`
	Project *ProjectSummary `json:"project"`

	Variables map[string]string `json:"vars"`
	State     TaskState         `json:"state"`
	Output    string            `json:"output"`
}

func (t *Task) Summary() *TaskSummary {
	return &TaskSummary{
		ID:       t.ID,
		Metadata: t.Metadata,
	}
}
